python/lsst/ts/hexrotcomm/command_telemetry_client.py

```diff
@@ -419,9 +417,60 @@ async def put_command(self, command):
-        # Note: this code allows command.counter to wrap around,
-        # without having to know the data type of that field.
-        command.counter = self._command_counts.get(command.code, 0) + 1
-        self._command_counts[command.code] = command.counter
-        await tcpip.write_from(self.command_writer, command)
+        async with self.command_lock:
+            command.counter = self._last_command.counter + 1
+            self._last_command = command
+            await tcpip.write_from(self.command_writer, command)
+
+            self.read_command_status_task = asyncio.create_task(
+                asyncio.wait_for(
+                    self.read_command_status(command.counter),
+                    timeout=COMMAND_STATUS_TIMEOUT,
+                )
+            )
+            await self.read_command_status_task
+            if self.command_status.status == CommandStatusCode.ACK:
+                return self.command_status.duration
+            elif self.command_status.status == CommandStatusCode.NO_ACK:
+                reason = self.command_status.reason.decode()
+                raise salobj.ExpectedError(reason)
+            else:
+                raise salobj.ExpectedError(
+                    f"Unknown command status {self.command_status.status}; "
+                    "low-level command assumed to have failed"
+                )
+
+    async def read_command_status(self, counter):
+        """Wait for a command status message whose counter matches
+        a specified value.
+        """
+        try:
+            while True:
+                if not self.command_connected:
+                    raise RuntimeError("Command socket disconnected during the command")
+
+                await tcpip.read_into(self.command_reader, self.command_status_header)
+                if (
+                    self.command_status_header.frame_id
+                    != structs.CommandStatus.FRAME_ID
+                ):
+                    self.log.warning(
+                        f"Ignoring message with unexpected frame_id: "
+                        f"{self.command_status_header.frame_id} != {structs.CommandStatus.FRAME_ID}"
+                    )
+                    continue
+
+                if not self.command_connected:
+                    raise RuntimeError("Command socket disconnected during the command")
+                await tcpip.read_into(self.command_reader, self.command_status)
+
+                if self.command_status.counter != counter:
+                    self.log.warning(
+                        "Ignoring command status for wrong command; "
+                        f"status.counter={self.command_status.counter} != expected value {counter}"
+                    )
+                    continue
+                return
+        except ConnectionError:
+            self.log.info("Command reader disconnected; closing client")
+            await self.basic_close()
```

Comment:
Do we want to consider the decode error here when receiving the command status message?